libdrgn/program.c

```diff
@@ -676,33 +703,31 @@ static struct drgn_error *drgn_program_cache_prstatus(struct drgn_program *prog)
 		       (offset = gelf_getnote(data, offset, &nhdr, &name_offset,
 					      &desc_offset))) {
 			const char *name;
-			uint32_t pr_pid;
-			struct drgn_prstatus_map_entry entry;
+			struct drgn_error *err;
 
 			name = (char *)data->d_buf + name_offset;
 			if (strncmp(name, "CORE", nhdr.n_namesz) != 0 ||
-			    nhdr.n_type != NT_PRSTATUS ||
-			    nhdr.n_descsz < pr_pid_offset + sizeof(pr_pid))
-				continue;
-			memcpy(&pr_pid,
-			       (char *)data->d_buf + desc_offset + pr_pid_offset,
-			       sizeof(pr_pid));
-			if (bswap)
-				pr_pid = bswap_32(pr_pid);
-			if (!pr_pid)
+			    nhdr.n_type != NT_PRSTATUS)
 				continue;
 
-			entry.key = pr_pid;
-			entry.value.str = (char *)data->d_buf + desc_offset;
-			entry.value.len = nhdr.n_descsz;
-			if (drgn_prstatus_map_insert(&prog->prstatus_cache,
-						     &entry, NULL) == -1)
-				return &drgn_enomem;
+			err = cb_fn(prog, (char *)data->d_buf + desc_offset, nhdr.n_descsz);
+			if (err)
+				return err;
 		}
 	}
 	return NULL;
 }
 
+static struct drgn_error *drgn_program_cache_prstatus(struct drgn_program *prog)
+{
+#ifdef WITH_LIBKDUMPFILE
+	if (prog->kdump_ctx) {
+		return drgn_program_prstatus_kdump_iter(prog, drgn_program_cache_prstatus_entry);
+	}
+#endif
+	return drgn_program_prstatus_elf_iter(prog, drgn_program_cache_prstatus_entry);
```

Comment:
This nicely avoids the issue of exporting `drgn_prstatus_map_insert()`. As a further improvement, it looks like `drgn_program_cache_prstatus_entry()` is the only thing passed as the callback, so it doesn't need to be a callback. How about we rename `drgn_program_prstatus_kdump_iter()` -> `drgn_program_cache_prstatus_kdump()`, fold `drgn_program_prstatus_elf_iter()` back into `drgn_program_cache_prstatus()`, and make them both call `drgn_program_cache_prstatus_entry()` directly?